Sources/Widgets/Widgets3D/ShapeWidget/example/index.js

```diff
@@ -276,10 +277,52 @@ reader
       }
     );
 
+    const updateWidgetVisibility = (widget, slicePos, i, widgetIndex) => {
+      const widgetVisibility =
+        widget.getWidgetState().getPoint1Handle().getOrigin()[i] ===
+          slicePos[i] ||
+        !scene.widgetManager.getWidgets()[widgetIndex].getPoint1();
+      widget.setVisibility(widgetVisibility);
+    };
+
+    const updateWidgetsVisibility = (position) => {
+      for (let i = 0; i < 3; i++) {
+        if (position[i] !== 0) {
+          /* testing if the widget is on the slice and has been placed to modify visibility */
+          updateWidgetVisibility(widgets.rectangleWidget, position, i, 0);
+          updateWidgetVisibility(widgets.ellipseWidget, position, i, 1);
+          updateWidgetVisibility(widgets.circleWidget, position, i, 2);
+        }
+      }
+    };
+
     const update = () => {
       const slicingMode = image.imageMapper.getSlicingMode() % 3;
 
       if (slicingMode > -1) {
+        const ijk = [0, 0, 0];
+        const position = [0, 0, 0];
+        const normal = [0, 0, 0];
+
+        // position
+        ijk[slicingMode] = image.imageMapper.getSlice();
+        data.indexToWorldVec3(ijk, position);
+
+        // circle/slice normal
+        ijk[slicingMode] = 1;
+        data.indexToWorldVec3(ijk, normal);
+        vtkMath.subtract(normal, data.getOrigin(), normal);
+        vtkMath.normalize(normal);
+
+        widgets.rectangleWidget.getManipulator().setOrigin(position);
+        widgets.rectangleWidget.getManipulator().setNormal(normal);
+        widgets.ellipseWidget.getManipulator().setOrigin(position);
+        widgets.ellipseWidget.getManipulator().setNormal(normal);
+        widgets.circleWidget.getManipulator().setOrigin(position);
+        widgets.circleWidget.getManipulator().setNormal(normal);
```

Comment:
I don't understand why this is necessary. The widget behavior itself should be setting the normal based on the camera, right? Can you explain?

Reply:
This was necessary to record the slice change. The normal should change when the camera moves position but the position of the manipulator should be updated when the user changes slice.

Reply:
> The normal should change when the camera moves position

So no need for `setNormal` here, only `setOrigin`?

> but the position of the manipulator should be updated when the user changes slice.

That sounds to me that the widget behavior should be able to do automatically, and that the application level shouldn't need to do that... I actually thought it did/should already, but clearly, it's not working. Would you agree?

___

_Note: I appreciate your patience in taking the time to explain this to me @ocrossi 👍_

Reply:
> So no need for setNormal here, only setOrigin?

My guess is that you need to set the normal too, especially if you switch the view type (I, J, K), isn't it ?

>That sounds to me that the widget behavior should be able to do automatically, and that the application level shouldn't need to do that... I actually thought it did/should already, but clearly, it's not working. Would you agree?

About not setting the origin at the application level, I think it's the opposite of what @floryst thinks [here ](https://github.com/Kitware/vtk-js/issues/1718#issuecomment-773486950). But that's definitely a question we need to solves for the next widgets :)

Reply:
> My guess is that you need to set the normal too, especially if you switch the view type (I, J, K), isn't it?

I don't think so, it should be handled [here](https://github.com/Kitware/vtk-js/blob/72e12774ea9e2485fa115e9ad9dc15f3e11285ba/Sources/Widgets/Widgets3D/ShapeWidget/behavior.js#L388). Also, as we've noted in both examples, the orientation of those green shape widgets itself does not seem to be the problem.

> About not setting the origin at the application level, I think it's the opposite of what @floryst thinks here. But that's definitely a question we need to solves for the next widgets :)

Ahah thanks for pointing that out @laurennlam, I missed that while catching up with the issue. Looks like we need to have a heated argument with @floryst! 🤜 🤛 Joke aside, I have no preference at the moment on doing this at the application level vs in the library, but from my understanding, all those widgets have been set up to handle things internally until now (through the behavior and manipulator objects), so we'd have to make a bigger switch here. Maybe we should all meet, or I can meet with Forrest first and report back.

Reply:
In the meantime @ocrossi I see you've force pushed some changes: did you test without those `setNormal` in the application `update`, since we believed [this one](https://github.com/Kitware/vtk-js/pull/1717/files#diff-09c501501761db156d6f7d2d7b93d812c1fccc886aac5a921dbda49b32481362R382) should handle it?

Reply:
It works fine without the setNormal from example. The useless lines are removed